### PR TITLE
refactor: read only txn request will use read state mechanism

### PR DESCRIPTION
**Pull request by themanforfree**
Reopen #367 

(Accidentally closed)

### CODE REVIEW

File: xlineapi/src/lib.rs [hunk 1]
```diff
@@ -671,6 +671,23 @@ impl TxnRequest {
         self.success.iter().all(read_only_checker) && self.failure.iter().all(read_only_checker)
     }
 
+    /// Checks whether a given `TxnRequest` is serializable or not.
+    pub fn is_serializable(&self) -> bool {
```
Comment (Phoenix500526) [inline xlineapi/src/lib.rs:675]: Please add a test to cover this method.

Comment (themanforfree) [reply xlineapi/src/lib.rs:675]: Fixed

File: xlineapi/src/lib.rs [hunk 2]
```diff
@@ -781,4 +781,82 @@ mod test {
...
+            compare: vec![],
+            success: vec![RequestOp {
+                request: serializable_req.clone(),
+            }],
+            failure: vec![RequestOp {
+                request: serializable_req.clone(),
+            }],
+        };
+
+        assert!(serializable_txn_req.is_serializable());
```
Comment (Phoenix500526) [inline xlineapi/src/lib.rs:802]: We should cover the happy path and the sad path here. Given that `is_serializable` is a recursive implementation, we'd better create a dedicated unit test in order to cover all possible branches.

**Review by Phoenix500526 — DISMISSED** [review]

File: xline/src/server/kv_server.rs [hunk 3]
```diff
@@ -310,19 +309,34 @@ where
...
+            let token = get_token(request.metadata());
+            let wrapper = RequestWithToken::new_with_token(request.into_inner().into(), token);
+            let propose_id = self.generate_propose_id();
+            let cmd = command_from_request_wrapper::<S>(propose_id, wrapper, None);
+            if !is_serializable {
+                self.wait_read_state(&cmd).await?;
+            }
+            self.do_serializable(cmd.request())?
+        } else {
+            let is_fast_path = false; // lock need revision of txn
```
Comment (bsbds) [inline xline/src/server/kv_server.rs:325]: after #391 the revision will be returned in execution result, so we can use fast path here for better performance

Comment (themanforfree) [reply xline/src/server/kv_server.rs]: Fixed

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by bsbds — APPROVED** [review]

**Pull request merged**